File: android/src/main/kotlin/com/debuggersight/hubspot_mobile_chat/HubspotMobileChatPlugin.kt
package com.debuggersight.hubspot_mobile_chat

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.hubspot.mobilesdk.HubspotManager
import com.hubspot.mobilesdk.HubspotWebActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** Error code shared with the Dart facade for the uniform error contract (FR-013). */
private const val CONFIG_ERROR_CODE = "hubspot_config_error"

/**
 * Android implementation of [HubspotHostApi]. Forwards every call to
 * [HubspotManager] (package `com.hubspot.mobilesdk`, SDK 1.0.8). No business
 * logic beyond marshalling and activity wiring.
 */
class HubspotMobileChatPlugin :
    FlutterPlugin,
    ActivityAware,
    HubspotHostApi {

    private lateinit var context: Context
    private var activity: Activity? = null
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        HubspotHostApi.setUp(binding.binaryMessenger, this)
        pushDelegate = HubspotFlutterApi(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HubspotHostApi.setUp(binding.binaryMessenger, null)
        pushDelegate = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    private fun manager(): HubspotManager = HubspotManager.getInstance(context)

    override fun configure(callback: (Result<Unit>) -> Unit) {
        // HubspotConfigError is internal to the SDK, so catch broadly and map to the
        // uniform config error code (FR-013).
        try {
            manager().configure()
            callback(Result.success(Unit))
        } catch (e: Throwable) {
            callback(Result.failure(FlutterError(CONFIG_ERROR_CODE, e.message, null)))
        }
    }

    override fun setUserIdentity(
        email: String,
        identityToken: String,
        callback: (Result<Unit>) -> Unit,
    ) {
        manager().setUserIdentity(email, identityToken)
        callback(Result.success(Unit))
    }

    override fun setChatProperties(
        properties: Map<String, String>,
        callback: (Result<Unit>) -> Unit,
    ) {
        manager().setChatProperties(properties)
        callback(Result.success(Unit))
    }

    override fun openChat(
        chatFlow: String?,
        pushData: PushData?,
        callback: (Result<Unit>) -> Unit,
    ) {
        val host = activity
        if (host == null) {
            callback(
                Result.failure(
                    FlutterError(CONFIG_ERROR_CODE, "No Activity to present chat", null),
                ),
            )
            return
        }
        // HubspotWebActivity reads "chatflow" (lowercase) + "hsPushData" extras.
        val intent = Intent(host, HubspotWebActivity::class.java)
        chatFlow?.let { intent.putExtra("chatflow", it) }
        host.startActivity(intent)
        callback(Result.success(Unit))
    }

    override fun registerPushToken(token: String, callback: (Result<Unit>) -> Unit) {
        // setPushToken is a suspend function; run it on a coroutine.
        scope.launch {
            try {
                manager().setPushToken(token)
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(FlutterError(CONFIG_ERROR_CODE, e.message, null)))
            }
        }
    }

    override fun logout(callback: (Result<Unit>) -> Unit) {
        // logout is a suspend function.
        scope.launch {
            try {
                manager().logout()
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(FlutterError(CONFIG_ERROR_CODE, e.message, null)))
            }
        }
    }

    companion object {
        /** Set while the engine is attached so the push service can reach Dart. */
        @Volatile
        var pushDelegate: HubspotFlutterApi? = null
    }
}

File: ios/hubspot_mobile_chat/Sources/hubspot_mobile_chat/HubspotMobileChatPlugin.swift
import Flutter
import HubspotMobileSDK
import SwiftUI
import UIKit

/// Error code shared with the Dart facade for the uniform error contract (FR-013).
private let configErrorCode = "hubspot_config_error"

/// iOS implementation of `HubspotHostApi`. Forwards every call to
/// `HubspotManager`; presents `HubspotChatView` for open-chat. No business
/// logic beyond marshalling and presentation.
///
/// Note: the generated `HubspotHostApi` protocol and its types (`PushData`,
/// etc.) are internal to this module, so the conformance methods are internal —
/// only `register(with:)` needs to be public for the plugin registrant.
public class HubspotMobileChatPlugin: NSObject, FlutterPlugin, HubspotHostApi {

  private var flutterApi: HubspotFlutterApi?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = HubspotMobileChatPlugin()
    HubspotHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    instance.flutterApi = HubspotFlutterApi(binaryMessenger: registrar.messenger())
    PushBridge.shared.flutterApi = instance.flutterApi
  }

  func configure(completion: @escaping (Result<Void, Error>) -> Void) {
    DispatchQueue.main.async {
      do {
        try HubspotManager.configure()
        completion(.success(()))
      } catch {
        completion(.failure(
          PigeonError(code: configErrorCode, message: error.localizedDescription, details: nil)))
      }
    }
  }

  func setUserIdentity(
    email: String, identityToken: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    DispatchQueue.main.async {
      HubspotManager.shared.setUserIdentity(identityToken: identityToken, email: email)
      completion(.success(()))
    }
  }

  func setChatProperties(
    properties: [String: String],
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    DispatchQueue.main.async {
      HubspotManager.shared.setChatProperties(data: properties)
      completion(.success(()))
    }
  }

  func openChat(
    chatFlow: String?, pushData: PushData?,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    DispatchQueue.main.async {
      guard let top = Self.topViewController() else {
        completion(.failure(
          PigeonError(code: configErrorCode, message: "No view controller to present chat",
            details: nil)))
        return
      }
      let chatView = HubspotChatView(manager: HubspotManager.shared, chatFlow: chatFlow)
      let host = UIHostingController(rootView: chatView)
      top.present(host, animated: true) { completion(.success(())) }
    }
  }

  func registerPushToken(
    token: String, completion: @escaping (Result<Void, Error>) -> Void
  ) {
    // App owns APNs registration (FR-007); forward the token to the SDK. The SDK
    // expects raw APNs token bytes, so decode the hex string the app provides.
    DispatchQueue.main.async {
      HubspotManager.shared.setPushToken(apnsPushToken: Self.dataFromHex(token))
      completion(.success(()))
    }
  }

  func logout(completion: @escaping (Result<Void, Error>) -> Void) {
    DispatchQueue.main.async {
      HubspotManager.shared.clearUserData()
      completion(.success(()))
    }
  }

  /// Decode an APNs device-token hex string ("a1b2…") into raw bytes.
  private static func dataFromHex(_ hex: String) -> Data {
    var data = Data(capacity: hex.count / 2)
    var index = hex.startIndex
    while index < hex.endIndex {
      let next = hex.index(index, offsetBy: 2, limitedBy: hex.endIndex) ?? hex.endIndex
      if let byte = UInt8(hex[index..<next], radix: 16) {
        data.append(byte)
      }
      index = next
    }
    return data
  }

  @MainActor
  private static func topViewController(
    _ base: UIViewController? = UIApplication.shared.connectedScenes
      .compactMap { ($0 as? UIWindowScene)?.keyWindow }
      .first?.rootViewController
  ) -> UIViewController? {
    if let nav = base as? UINavigationController {
      return topViewController(nav.visibleViewController)
    }
    if let tab = base as? UITabBarController {
      return topViewController(tab.selectedViewController)
    }
    if let presented = base?.presentedViewController {
      return topViewController(presented)
    }
    return base
  }
}

/// Bridges incoming APNs payloads (forwarded by the app) to Dart (FR-007).
public class PushBridge: NSObject {
  public static let shared = PushBridge()
  var flutterApi: HubspotFlutterApi?

  /// Call from the app's `didReceiveRemoteNotification` for HubSpot chat pushes.
  public func handleRemoteNotification(_ userInfo: [AnyHashable: Any]) {
    var raw = [String?: String?]()
    for (key, value) in userInfo {
      if let key = key as? String { raw[key] = "\(value)" }
    }
    let pushData = PushData(
      messageId: userInfo["messageId"] as? String,
      threadId: userInfo["threadId"] as? String,
      raw: raw)
    flutterApi?.onNewMessagePush(data: pushData) { _ in }
  }
}
